core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -535,7 +536,7 @@ class GroupMetadataManager(val brokerId: Int,
 
     val numExpiredOffsetsRemoved = inWriteLock(offsetExpireLock) {
       val expiredOffsets = offsetsCache.filter { case (groupTopicPartition, offsetAndMetadata) =>
-        offsetAndMetadata.expireTimestamp < startMs
+        offsetAndMetadata.expireTimestamp < startMs || !metadataCache.contains(groupTopicPartition.topicPartition.topic)
```

Comment:
Is there any concern about increased contention for the partitionMetadata lock while this is running? I wonder if it would make sense to add a topics() method to MetadataCache which returns a copy of the cached topics. Then you could use the copy and avoid the lock.